vibin/mediaservers/cxnv2.py

```diff
@@ -0,0 +1,589 @@
+        Artists`.
+        """
+        tracks = self.tracks_by_album_id[album_id]
+        if not tracks:
+            return "Unknown"
+        (artist, count) = Counter([t.artist for t in tracks]).most_common(1)[0]
+        if artist and count > len(tracks) / 2:
+            return artist
+        else:
+            return "Various Artists"
+
+    def _guess_genre(self, album_id: MediaId):
+        """Infers a genre from an album's tracks.
+
+        If most of the tracks share a genre, uses that genre. Otherwise, return
+        the string 'Unknown' which the UI expects in the absense of a genre.
+        """
+        tracks = self.tracks_by_album_id[album_id]
+        if not tracks:
+            return "Unknown"
+        (genre, count) = Counter([t.genre for t in tracks]).most_common(1)[0]
+        if genre and count > len(tracks) / 2:
+            return genre
+        else:
+            return "Unknown"
+
+    def create_media_folder(self, result: _BrowseResult) -> MediaFolder:
+        """Creates a new MediaFolder with information from this _BrowseResult."""
+        return MediaFolder(
+            title = result.title,
+            album_art_uri = result.album_art_uri,
+            artist = result.artist,
+            genre = result.genre,
+        )
+
+    def create_model_object(self, result: _BrowseResult) -> Track | Album | MediaFolder:
+        """Returns a model object for this _BrowseResult."""
+        match result.type:
+            case "object.item.audioItem.musicTrack":
+                return self.create_track(result)
+            case "object.container.album.musicAlbum":
+                return self.create_album(result)
+            case _:
+                return self.create_media_folder(result)
+
+    def create_artist(self, name: str):
+        """Creates an Artist object given a name.
+
+        Ids are synthetic and based simply on the name.
+        """
+        return Artist(
+            id = base64.urlsafe_b64encode(name.encode()),
+            title = name,
+            genre = "Unknown",
+            album_art_uri = None
+        )
+
+    @staticmethod
+    def _group_tracks_by_album_id(tracks: list[Track]) -> dict[MediaId, list[Track]]:
+        out = {}
+        for track in tracks:
+            if track.albumId not in out: out[track.albumId] = []
+            out[track.albumId].append(track)
+
+        return out
+
+    @staticmethod
+    def stabilize_resource_uri(resource: str):
+        """Strips the unstable portion of a resource path, leaving a portion
+        which survives standby mode and is still likely to be a unique
+        identifier for a track.
+
+        Resource paths are in the form:
+        `file:///tmp/usm/27/music/Kosheen/Resist/01%20-%20Demonstrate.mp3`
+        ...where the number is unstable. Returns the rest of the path.
+        """
+        return '/'.join(urlparse(resource).path.split('/')[4:])
+
+
+class CXNv2(MediaServer):
```

Comment:
Do you know if this might work for other Cambridge Audio devices? If so then it could be named something not device specific, like `CambridgeAudioUSB`. I'm happy to keep it named as-is though, especially if you're unsure.

Another option might be `CXNv2USB` since it's technically a USB MediaServer implementation, and one day other Cambridge-specific MediaServer abstraction layers might make sense to implement too (for services like Tidal, Qobuz, etc) -- although I really haven't thought that through, and streaming service support isn't currently on the todo list anyway.

Reply:
I've only got my own CXNv2 to play with - it'd be nice if this worked with the CXN100 but I've no idea.

I've gone for CXNv2USB.